### DISCUSSION TITLE
Xpra X11 missing adobe courier fonts

### DISCUSSION THREAD

**Discussion by Simba98** [Xpra-org/xpra#3924] [Q&A] [answered]
After install x11-fonts package in Ubuntu
```
xlsfonts | grep adobe

Output: many font such as -adobe-courier-bold-r-normal--0-0-75-75-m-0-iso8859-10
```
But if we use xpra to connect.
```
xlsfonts | grep adobe

Outputs nothing
```

**Comment by totaam** [comment, marked as answer]
Perhaps `Xvfb` doesn't load the same font directories as the real Xorg.
Switching to [Xdummy](https://github.com/Xpra-org/xpra/blob/master/docs/Usage/Xdummy.md) would probably resolve this issue.

If you really have to stick with Ubuntu, this may be more tricky. YMMV.

**Reply by Simba98** [reply]
I solved this issue by reinstall xfonts-100dpi ( Other uses may try xorg-x11-fonts xfonts-base xfonts-100dpi xfonts-75dpi xfonts-cyrillic, etc)
And in Xpra, the fonts are loaded by X11 server, which is usually launched by Xpra server. This may be common knowledge of experts, but I do not know it and found it by some testing.

The following figure shows the difference between Xpra and normal X11 C/S. 
The first row is Xpra and the second is X11 C/S.

<img width="686" alt="image" src="https://github.com/Xpra-org/xpra/assets/7941967/5641317f-74e3-4565-a3a1-5ca93bec2b74">

**Reply by totaam** [reply]
That diagram isn't very clear but it looks completely wrong to me.
On the client side, xpra doesn't necessarily talk to an X11 server (it can run on MS Windows, MacOS, Wayland, etc) and definitely does not use a "fonts provider" at all.
There is no font rendering at all client side.

**Reply by Simba98** [reply]
The second row is the normal X11 Server/Client model, and the X11 Server provides fonts. So if the GUI is shown in MS Windows, the MS Win program provides the fonts.

But in the first row, for Xpra, the Xpra Client did not provide the font; the X11 Server, launched by Xpra Server, provides fonts; so I have to install the fonts on the X11/Xpra Server side (the middle machine), instead of installing the font on the GUI showing machine (the left machine, Xpra Client side).

Thank you for your time, and appreciate correcting me if I misunderstood it.

**Reply by totaam** [reply]
The xpra server does **NOT** provide fonts at all, it knows nothing about fonts.
That's a job for X11.

**Reply by Simba98** [reply]
OK, so the fonts are provided by the X11 server, which usually is started by the Xpra server in the same machine.
Is that correct? And I edited the above reply based on your correcting.

**Reply by totaam** [reply]
See [Understanding where X11 fonts come from](https://superuser.com/a/1324263/58851).

Nowadays, most applications use their own rendering (ie: GTK does this via [pango](https://pango.gnome.org/)) and only render pixels to the X11 server.

Also note that although X11 client applications usually reside on the same host as the X11 server (and this is very strongly recommended with xpra), this does not have to be the case. And they can both use the same fonts.

**Comment by Simba98** [comment]
> Perhaps `Xvfb` doesn't load the same font directories as the real Xorg. Switching to [Xdummy](https://github.com/Xpra-org/xpra/blob/master/docs/Usage/Xdummy.md) would probably resolve this issue.
> 
> If you really have to stick with Ubuntu, this may be more tricky. YMMV.

Ok, I will test if this is due to Xvfb. The host server must be Ubuntu (or CentOS 7, which is even more incompatible), I can run docker for Xpra, so maybe I can try other distributions in docker to check if this issue is only happening in Ubuntu.
